src/text.c

```diff
@@ -0,0 +1,358 @@
+ *
+ *   somekey=somevalue
+ *
+ * To distinguish these from path lines, if there is a leading '/' in the key, it must be escaped
+ * (slashes later in the line don't have to be escaped):
+ *
+ *   \/slashedkey=somevalue
+ *
+ * Other than that, at any postion, '\\' and '\n' must be escaped:
+ *   
+ *   some\\key=some\nvalue
+ *
+ * In the key, '=' must also be escaped (not necessary in the value):
+ *
+ *   some\=key=some=value
+ * 
+ * --------
+ * Example:
+ *
+ *   /
+ *   key=intheroot
+ *   \/slashedkey=somevalue
+ *   some\\key=some\nvalue
+ *   some\=key=some=value
+ *
+ *   /subns
+ *   some=stuff in the sub-namespace
+ *
+ *   /subns/deeper
+ *   this=is in /subns/deeper
+ *
+*/
+
+static int cmp_ns(const void *a, const void *b) {
+	const SdbNs *nsa = a;
+	const SdbNs *cia = b;
+	return strcmp (nsa->name, cia->name);
+}
+
+static bool text_fsave(Sdb *s, FILE *f, bool sort, SdbList *path) {
+// n = position we are currently looking at
+// p = position until we have already written everything
+// macro to flush a block of text that doesn't have to be escaped
+#define FLUSH do { if (p != n) { fwrite (p, 1, n - p, f); p = n; } } while (0)
+// macro to flush and skip a char
+#define SKIP do { FLUSH; p++; } while (0)
+	// path
+	fwrite ("/", 1, 1, f); // always print a /, even if path is empty
+	SdbListIter *it;
+	const char *path_token;
+	bool first = true;
+	ls_foreach (path, it, path_token) {
+		if (first) {
+			first = false;
+		} else {
+			fwrite ("/", 1, 1, f);
+		}
+		// write and escape the path
+		const char *p = path_token;
+		const char *n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '/':
+				SKIP;
+				fwrite ("\\/", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+	}
+	fwrite ("\n", 1, 1, f);
+
+	// k=v entries
+	SdbList *l = sdb_foreach_list (s, sort);
+	SdbKv *kv;
+	ls_foreach (l, it, kv) {
+		const char *k = sdbkv_key (kv);
+		const char *v = sdbkv_value (kv);
+		// escape leading '/'
+		if (*k == '/') {
+			fwrite ("\\", 1, 1, f);
+		}
+
+		// write and escape key
+		const char *p = k;
+		const char *n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '=':
+				SKIP;
+				fwrite ("\\=", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+
+		fwrite ("=", 1, 1, f);
+
+		// write and escape value
+		p = v;
+		n = p;
+		while (*n) {
+			switch (*n) {
+			case '\\':
+				SKIP;
+				fwrite ("\\\\", 1, 2, f);
+				break;
+			case '\n':
+				SKIP;
+				fwrite ("\\n", 1, 2, f);
+				break;
+			}
+			n++;
+		}
+		FLUSH;
+		fwrite ("\n", 1, 1, f);
+	}
+	ls_free (l);
+
+	// sub-namespaces
+	l = s->ns;
+	if (sort) {
+		l = ls_clone (l);
+		ls_sort (l, cmp_ns);
+	}
+	SdbNs *ns;
+	ls_foreach (l, it, ns) {
+		fwrite ("\n", 1, 1, f);
+		ls_push (path, ns->name);
```

Comment:
mmm is it ever possible to have more than one name in `path` being passed to `text_fsave`? It seems to me that this way `text_fsave` will always have at most one path to deal with, so maybe instead of passing a SdbList we can just pass a single string.

Reply:
> mmm is it ever possible to have more than one name in path being passed to text_fsave?

Yes, because `text_fsave` recurses. When you have the path "/some/path/here", the list will be `["some", "path", "here"]`. This is to avoid clashing with '/' chars in path tokens.

Reply:
I see now yeah :+1: